### PR TITLE
Limit connections

### PR DESCRIPTION
**Pull request by Furisto**
## Description
Introduce limiting of new connections created per minute. For more info see the [RFC](https://www.notion.so/gitpod/Rate-limit-network-connections-b3ac59e7db5d41198d63cc7f374c0333).

## Related Issue(s)
Related to https://github.com/gitpod-io/gitpod/issues/10651

## How to test
- Open a workspace in the preview environment
- Annotate the pod with gitpd.io/netConnLimit=true
- This should add additonal nftable rules to the network namespace of the pod (ssh into the node and then use `nsenter -t workspacekitPid -n` and `nft list ruleset`)
- Metrics can be obtained with `kubectl port-forward ds/ws-daemon 9500:9500,` then get the metrics with `curl -XGET localhost:9500/metrics`
- Run roblox scanner

## Release Notes

```release-note
Limit the rate at which network connections can be made by a workspace 
```
## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/ws-daemon/nsinsider/main.go
Comment (Furisto) [reply components/ws-daemon/nsinsider/main.go:572]: Can you explain what kind of issues have been observed in the context of kubelet? I was not able to find anything.

**Review by MrSimonEmms — APPROVED** [review]

**Review by sagor999 — APPROVED** [review]
LGTM
adding hold in case if @aledbf comment requires resolving prior to merging
feel free to remove it.

/hold

File: components/ws-daemon/nsinsider/main.go
Comment (Furisto) [reply components/ws-daemon/nsinsider/main.go:572]: @aledbf Are you ok with merging this or would you like to discuss it further?

Comment (aledbf) [reply components/ws-daemon/nsinsider/main.go:572]: I am not sure we should merge this without knowing the behavior of different `nftable` versions and if there is any conflict with the CNI bandwidth plugin.

Did you test this change in an ephemeral cluster?

Comment (Furisto) [reply components/ws-daemon/nsinsider/main.go:572]: workspace preview

Comment (kylos101) [reply components/ws-daemon/nsinsider/main.go:572]: We used nftables to replace slirp4net, if there is an issue with versions, would we have more fundamental issues with workspace networking? 
* This makes me wonder if we should add related [runtime tests for networking](https://www.notion.so/gitpod/Common-runtime-tests-4e3b05c3b81e4ff6810572aee66ffad3). 
* I ask because I'm not sure what dimensions to consider.

@aledbf I like the idea of testing with an ephemeral cluster. Can you [add a related test here](https://www.notion.so/gitpod/Common-runtime-tests-4e3b05c3b81e4ff6810572aee66ffad3), so we can have a repeatable way to do in the future? 

@Furisto let's leave on-hold till we've had a chance to test in an ephemeral cluster (to test for impacts to CNI bandwidth plugin).

Comment (utam0k) [inline components/ws-daemon/nsinsider/main.go:572]: Since this function runs before `pivot_root`, I believe it is safe to specify the `nftable` bin with the complete path at this stage. Am I right? If yes, please add a comment that why it is safe to run.

Comment (utam0k) [reply components/ws-daemon/nsinsider/main.go:572]: I'm concerned that the user may be able to override this bin with a custom image.

File: components/ws-daemon/nsinsider/main.go
Comment (utam0k) [inline components/ws-daemon/nsinsider/main.go:494]: I think it would be better to move ip-fowarding rules to this table. WDYT? Maybe it is good for another PR to take care of that.

File: components/ws-daemon/nsinsider/main.go
Comment (utam0k) [reply components/ws-daemon/nsinsider/main.go:572]: cc: @mrzarquon 
We should install the nftables command, right? If yes, we have to change the image and notify the self-hosted team.

File: components/ws-daemon/nsinsider/main.go
Comment (utam0k) [inline components/ws-daemon/nsinsider/main.go:569]: Can we use set syntax instead of meter as official documentation says here?
https://wiki.nftables.org/wiki-nftables/index.php/Meters
> Note that the meter keyword is obsolete, the dynamic set and map syntax is now preferred for consistency.

File: components/ws-daemon/nsinsider/main.go
Comment (Furisto) [reply components/ws-daemon/nsinsider/main.go:572]: We do not enter the mount namespace of the workspace, only the network namespace. The binary [is part of ws-daemon](https://github.com/gitpod-io/gitpod/blob/f4ef8c3457b3bd352554c3feb76cb8f1bcc373c1/components/ws-daemon/leeway.Dockerfile#L25).

File: components/ws-daemon/nsinsider/main.go
Comment (Furisto) [reply components/ws-daemon/nsinsider/main.go:494]: Yes, I have thought about encapsulating all the network rules we have within the gitpod table, but I felt that for this PR it is out of scope.

File: components/ws-daemon/nsinsider/main.go
Comment (Furisto) [reply components/ws-daemon/nsinsider/main.go:569]: I think we can. Thanks for the hint. I also will update this rule to use a named counter, so that we can scrape metrics about dropped packets more easily.

File: components/ws-daemon/nsinsider/main.go
Comment (utam0k) [reply components/ws-daemon/nsinsider/main.go:494]: 🆗

File: components/ws-daemon/nsinsider/main.go
Comment (utam0k) [reply components/ws-daemon/nsinsider/main.go:572]: 💯 Can I ask you to add a comment about that? Why? I think this is a very important security point that is difficult to see from the code.

**Comment by Furisto** [comment]
- Limiting is now setup by ws-daemon instead of workspacekit
- We use an annotation (gitpod.io/netConnLimit) to detect workspaces that should be limited
- The go library for nftables recently gained support for dynsets, so using nft is not necessary anymore
- Metrics for dropped bytes/packets have been added
- The network limiting annotation is now set by ws-manager if the WORKSPACE_CONNECTION_LIMITING feature is send as part of the request. The decision to send the flag will be made based on [the plan that the user is on](https://github.com/gitpod-io/gitpod/pull/12162). This part is not yet fully implemented because we do need to deploy this PR first, so that the flag can be handled by ws-manager.

**Comment by kylos101** [comment]
:wave: @Furisto I've added this to our project for team visibility, and changed the description from "fixes" to "related to" (assuming merging this PR won't necessarily close the epic).

Will this audit connections (which we would have limited), or, will it actually limit? I assume audit, but wanted to double check.

**Comment by Furisto** [comment]
@kylos101 
> Will this audit connections (which we would have limited), or, will it actually limit? I assume audit, but wanted to double check.

It will only [audit](https://github.com/gitpod-io/gitpod/blob/a87d9e0b246092f78c2b3113b6776a6f5e50e535/components/ws-daemon/nsinsider/main.go#L616) them

File: components/common-go/kubernetes/kubernetes.go
Comment (utam0k) [inline components/common-go/kubernetes/kubernetes.go:74]: I felt it would be good that the unit is included in the annotation name.
```suggestion
	WorkspaceNetConnLimitAnnotation = "gitpod.io/netConnLimitPerMin"
```

File: components/ws-daemon/pkg/nsinsider/nsinsider.go
Comment (utam0k) [inline components/ws-daemon/pkg/nsinsider/nsinsider.go:1]: 👍 Good refactoring. I love these code.

File: components/ws-daemon/pkg/netlimit/netlimit.go
Comment (utam0k) [inline components/ws-daemon/pkg/netlimit/netlimit.go:28]: How about using `sync.Map` instead of `mu`? And we can use `struct{}{}` as mark instead of `bool`
```suggestion
	limited        sync.Map
```

**Review by utam0k — CHANGES_REQUESTED** [review]
Almost are good. I left some comments, please check 🙏

**Review by geropl — APPROVED** [review]
Reviewed ws-manager-api per code owners: change LGTM

File: components/ws-daemon/pkg/netlimit/netlimit.go
Comment (Furisto) [reply components/ws-daemon/pkg/netlimit/netlimit.go]: Using a sync.Map would introduce a race condition because we could get a WorkspaceAdded event immediately followed by a WorkspaceUpdated event. If WorkspaceAdded is still processed then the sync.Map would not contain an entry yet and we would try to limit twice. Using a mutex serializes the updates to the limiting rules. I have updated the map to use an empty struct.

**Comment by Furisto** [comment]
@utam0k PTAL

File: components/ws-daemon/pkg/netlimit/netlimit.go
Comment (utam0k) [reply components/ws-daemon/pkg/netlimit/netlimit.go]: AFAI, in that case, you said either event should be executed. What about doing a LoadOrStore first and Delete if it fails?　　e.g. https://github.com/gitpod-io/gitpod/pull/12215#discussion_r950433725
This is more efficient because the lock is not applied to the entire map.
I learned `sync.Map` from this article.
https://sreramk.medium.com/go-inside-sync-map-how-does-sync-map-work-internally-97e87b8e6bf

Comment (Furisto) [reply components/ws-daemon/pkg/netlimit/netlimit.go]: The way I see it, this is not an operation that is called many thousands of times per second, but only rarely (at most 3 times per second in my experience, usually less). Any contention on the lock is likely to be short lived. I consider simplicity therefore more important than efficiency in this case.

Comment (utam0k) [reply components/ws-daemon/pkg/netlimit/netlimit.go]: Make sense🙆

**Review by utam0k — APPROVED** [review]
Awesome

**Pull request merged**
